sidecar/internal/service/reclaimspace.go

```diff
@@ -61,22 +69,47 @@ func (rs *ReclaimSpaceServer) ControllerReclaimSpace(
 	pvName := req.GetPvName()
 	klog.Info(pvName)
 
-	// TODO: get following params using kubeClient and pvName
+	pv, err := rs.kubeClient.CoreV1().PersistentVolumes().Get(ctx, pvName, metav1.GetOptions{})
+	if err != nil {
+		klog.Errorf("Failed to get pv: %v", err)
+		return nil, status.Errorf(codes.InvalidArgument, "failed to get pv %q", pvName)
+	}
+
+	if pv.Spec.CSI == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "pv %q is not a CSI volume", pvName)
+	}
+
+	volID := pv.Spec.CSI.VolumeHandle
+	volAttributes := pv.Spec.CSI.VolumeAttributes
 
 	csiReq := &csiReclaimSpace.ControllerReclaimSpaceRequest{
-		VolumeId:   "",
-		Parameters: map[string]string{},
-		Secrets:    map[string]string{},
+		VolumeId:   volID,
+		Parameters: volAttributes,
 	}
-	res, err := rs.controllerClient.ControllerReclaimSpace(ctx, csiReq)
+
+	// FIXME: use ControllerPublishSecret instead, but it is not set in the PV
+	if pv.Spec.CSI.NodeStageSecretRef != nil {
+		// Get the secrets from the k8s cluster
+		csiReq.Secrets, err = kube.GetSecret(ctx, rs.kubeClient, pv.Spec.CSI.NodeStageSecretRef.Name, pv.Spec.CSI.NodeStageSecretRef.Namespace)
+		if err != nil {
+			klog.Errorf("Failed to get secret: %v", err)
+			return nil, status.Error(codes.InvalidArgument, err.Error())
+		}
+	}
+
+	csiRes, err := rs.controllerClient.ControllerReclaimSpace(ctx, csiReq)
 	if err != nil {
 		return nil, err
 	}
+	res := &proto.ReclaimSpaceResponse{}
+	if csiRes.PreUsage != nil {
```

Comment:
csiRes can be nil, right? Do we need a check like this?

```
if csiRes !=nil {
    if csiRes.PreUsage != nil {
              res.PreUsage = &proto.StorageConsumption{UsageBytes: csiRes.PreUsage.UsageBytes}
	}
     if csiRes.PostUsage != nil {
		res.PostUsage = &proto.StorageConsumption{UsageBytes: csiRes.PostUsage.UsageBytes}
	}
}

Reply:
yes. This is to avoid panic. We can instead send the empty structure. csiRes can be nil as it is optional , it will not throw any error but will get stuck in panic state as shown by @Rakshith-R . @nixpanic let me know if this is correct.

Reply:
Indeed, if `csiRes` is `nil`, `err` should be set. Checking `csiRes == nil` in below the `if err != nil` and returning a suitable error would be good.

Reply:
Yes IMO too adding the err in case csiRes is nil will be better, as it won't get stuck in a panic state and also we can't return an empty struct.

Reply:
In general, readability of the code is best when following a flow like:
```go
res, err := someFunction()

// handle failure cases
if err != nil {
   return fmt.Errorf("someFunction failed: %w", err)
}
if res == nil {
    return fmt.Errorf("someFunction returned and unexpected result for %q", someStringArg)
}

// continue with success case
```

Reply:
above works for me. res nil check is suggested here https://github.com/csi-addons/kubernetes-csi-addons/pull/32#discussion_r775018426 hence the question